sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md

```diff
@@ -0,0 +1,731 @@
+# Explicit SIMD Programming Extension for DPC++
+
+## Introduction
+
+The main motivation for introducing the Explicit SIMD Programming (ESP) DPC++
+extension is enabling low-level efficient programming for Intel Gen
+architectures. More specifically, Explicit SIMD provides the following
+additional features:
+- Manual vectorization of device code using the `simd` class mapped to Gen's
+  general register file. This allows to write efficient code not relying on
+  further widening by the compiler, as with traditional SIMT programming.
+- Low-level APIs efficiently mapped to Gen architecture, such as block reads.
+
+One of the future directions is enabling this extension for other architectures,
+such as x86, with extracting and clearly marking generic and target-dependent
+API portions.
+
+## Explicit SIMD execution model
+
+Explicit SIMD execution model is a basically an equivalent of the base SYCL
+execution model with subgroup size restricted to 1. Which means in ESP
+it is guaranteed that there is a single work item in any workgroup, and it maps
+to a single hardware thread. All standard SYCL APIs continue to work,
+including `sycl::intel::sub_group` ones, which become either a no-op or
+trivial. E.g. a barrier becomes just a memory fence for a compiler, collectives
+just return the value in the single work-item. Another consequences of the unit
+subgroup size is guaranteed independent forward progress between work-items on
+some Gen architecture generations.
+
+## Explicit SIMD extension APIs
+
+Explicit SIMD APIs can be used only in code to be executed on Intel Gen
+architecture devices and the host device for now. Attempt to run such code on
+other devices will result in error. 
+
+Kernels and `SYCL_EXTERNAL` functions using ESP must be explicitly marked with
+the `EXPLICIT_SIMD` attribute macro. Subgroup size quiery within such
```

Comment:
To have flexibility in implementation. At this point (and likely in mid-term too) we need to distinguish SYCL `reqd_sub_group_size(1)` kernels from ESIMD kernels. So `EXPLICIT_SIMD` expands to `sycl_explicit_simd` attribute and then FE adds `reqd_sub_group_size(1)` to all kernels marked with `sycl_explicit_simd`.

Reply:
Why not ask people to put the `sycl_explicit_simd` attribute for now? When it isn't needed anymore we can make it a deprecated alias for `reqd_sub_group_size(1)`. The marco on the hand can't be deprecated. And in general we try to avoid macros.

Reply:
With macro we can just redefine EXPLICIT_SIMD to `reqd_sub_group_size(1)` later w/o any changes to the user code.

Reply:
Yes but then `EXPLICIT_SIMD` stays in the user code rather than gets a deprecation warning. That's a disadvantage not an advantage.

Reply:
OK, I will make the change.